src/ui/public/visualize/components/visualization_chart.tsx

```diff
@@ -0,0 +1,156 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import React from 'react';
+import * as Rx from 'rxjs';
+import { debounceTime, filter, share, switchMap, tap } from 'rxjs/operators';
+
+import { PersistedState } from '../../persisted_state';
+import { dispatchRenderComplete, dispatchRenderStart } from '../../render_complete';
+import { ResizeChecker } from '../../resize_checker';
+import { Vis, VisualizationController } from '../../vis';
+import { getUpdateStatus } from '../../vis/update_status';
+
+interface VisualizationChartProps {
+  onInit?: () => void;
+  uiState: PersistedState;
+  vis: Vis;
+  visData: any;
+}
+
+class VisualizationChart extends React.Component<VisualizationChartProps> {
+  private resizeChecker?: ResizeChecker;
+  private visualization?: VisualizationController;
+  private chartDiv = React.createRef<HTMLDivElement>();
+  private containerDiv = React.createRef<HTMLDivElement>();
+  private renderSubject: Rx.Subject<{
+    vis: Vis;
+    visData: any;
+    container: HTMLElement;
+  }>;
+  private renderSubscription: Rx.Subscription;
+
+  constructor(props: VisualizationChartProps) {
+    super(props);
+
+    this.renderSubject = new Rx.Subject();
+    const render$ = this.renderSubject.asObservable().pipe(share());
+
+    const success$ = render$.pipe(
+      tap(() => {
+        if (this.chartDiv.current) {
+          dispatchRenderStart(this.chartDiv.current);
+        }
+      }),
+      filter(
+        ({ vis, visData, container }) => vis && container && (!vis.type.requiresSearch || visData)
+      ),
+      debounceTime(100),
```

Comment:
I know that this was here before, but is there any more "pragmatic" way to avoid this?
Adding timers and debouncing usually increase the possibility to have flaky tests because we are not tied to a specific and clear state and we are saying: ok **maybe** there will be other 2 **or** 3   subsequent call/updates of the `renderSubject` one next to the other and it's better to wait **some** time.
I think this kind of uncertainty brings the flakyness to our test.
Any thoughts on that?

Reply:
The most common use-case at the moment is resizing the visualization, e.g. by resizing the window or resizing the dashboard panel it's in. In case we don't debounce, we are very laggy due to too many renders, while scaling a dashboard panel. Though it wouldn't be needed in all rendering triggering cases, but I am not sure if we actually simplify the logic and increase maintainability if we separate between renders, that need the debounce and some that doesn't.